### PR TITLE
fix: price update

### LINKED ISSUE #837: Trade Page: Prices and liquidity details do not automatically update
Right now, prices don't update unless the user changes the page or refreshes. This is not a good UX.

In the video, I had the page open for a long time. Then, I started recording, then I refreshed the page so you can see the price changed.

However, I just want to give a warning. Make sure that, after a user clicks "swap", we NEVER UPDATE minimum to receive. This must be fixed after the user clicks "swap". No updates between clicking "swap" and clicking "confirm". 

https://images.zenhubusercontent.com/322745209/5dabe9bd-b98c-43ca-9eae-4aa69e265d3b/screen_recording_2022_01_06_at_5_11_26_pm.mov

### PR DESCRIPTION
**Pull request by octocat512**
- fetch pairs info per block
- refactor useV2Pairs

close #837

### CODE REVIEW

**Comment by octocat512** [comment]
I'd like you to refactor the useQueuePair. @lijac

File: src/hooks/useV2Pairs.ts [hunk 1]
```diff
@@ -22,82 +23,91 @@ export enum PairState {
...
+                  stats = await bnJs.Dex.getPoolStats(poolId);
+                } catch (err) {
+                  return [PairState.NOT_EXISTS, null];
+                }
+
+                const baseReserve = new BigNumber(stats['base'], 16).toFixed();
+                const quoteReserve = new BigNumber(stats['quote'], 16).toFixed();
+                const totalSupply = new BigNumber(stats['total_supply'], 16).toFixed();
+
+                if (stats['base_token'] === tokenA.address) {
```
Comment (lijac) [inline src/hooks/useV2Pairs.ts:60]: Can we refactor this?

`const [reserveA, reserveB] = (stats['base_token'] === tokenA.address) ? [baseReserve, quoteReserve] : [quoteReserve, baseReserve];`

**Review by lijac — COMMENTED** [review]
Good work! I left a comment.

File: src/hooks/useV2Pairs.ts [hunk 2]
```diff
@@ -22,82 +23,91 @@ export enum PairState {
...
   }, [currencies]);
 
-  // const lastBlockNumber = useBlockNumber();
+  const [pairs, setPairs] = useState<[PairState, Pair | null][]>(Array(tokens.length).fill([PairState.LOADING, null]));
+
+  const lastBlockNumber = useBlockNumber();
 
   useEffect(() => {
-    setReserves(Array(tokens.length).fill(PairState.LOADING));
+    setPairs(Array(tokens.length).fill([PairState.LOADING, null]));
```
Comment (lijac) [inline src/hooks/useV2Pairs.ts:35]: Any reason to extract it as a separate hook?

Comment (octocat512) [reply src/hooks/useV2Pairs.ts:35]: Could you explain more?

Comment (octocat512) [reply src/hooks/useV2Pairs.ts:35]: The reason is to keep previous data. This is why I refactored the code. @lijac

**Comment by octocat512** [comment]
@lijac how's the liquidity details update?

**Comment by octocat512** [comment]
- [ ] use the `getPoolStatsForPair` function

**Comment by lijac** [comment]
- [ ] update pool (pair) reserves periodically

File: src/app/components/trade/LiquidityDetails.tsx [hunk 3]
```diff
@@ -22,7 +22,7 @@ import { DropdownPopper } from 'app/components/Popover';
 import { Typography } from 'app/theme';
 import bnJs from 'bnJs';
 import { BIGINT_ZERO, FRACTION_ONE, FRACTION_ZERO } from 'constants/misc';
-import { BalanceState, useAvailablePairs, useBalances } from 'hooks/useV2Pairs';
+import { BalanceData, useAvailablePairs, useBalances } from 'hooks/useV2Pairs';
```
Comment (octocat512) [inline src/app/components/trade/LiquidityDetails.tsx:25]: Why did you change the variable name?

Comment (octocat512) [reply src/app/components/trade/LiquidityDetails.tsx:25]: @lijac

**Comment by octocat512** [comment]
I think we can use the `useV2Pairs` hook instead of the `useReserves` hook in the `useAvailablePairs` hook. So we can remove the `useReserves` hook.

```
  const availableTokenPairs = availablePoolIds.map(el => [el.token0, el.token1]);
  const reserves = useV2Pairs();

```
I'd also like to change the data structure.

interface PoolIdData {
  poolId: number;
  tokens: [Token, Token];
}

File: src/app/components/trade/LiquidityDetails.tsx (hunk 3 above)
Comment (lijac) [reply src/app/components/trade/LiquidityDetails.tsx:25]: It is to ensure consistency, there are types (interfaces actually) `BalanceState`, `BalanceData`, `ReserveState`, `ReserveData`.

**Comment by lijac** [comment]
> 

Yes, we can use `useV2Pairs`.

About the data structure change, is there a specific reason?

**Comment by octocat512** [comment]
> > 
> 
> Yes, we can use `useV2Pairs`.
> 
> About the data structure change, is there a specific reason?

It is just for the convention. I think we used the tuple data structure in several parts. for example, `useV2Pairs`

**Comment by octocat512** [comment]
Please remove `useReserves` hook and set the 10 sec period. @lijac

File: src/hooks/useV2Pairs.ts [hunk 4]
```diff
@@ -22,91 +21,103 @@ export enum PairState {
...
-  >([]);
-
   const tokens = useMemo(() => {
     return currencies.map(([currencyA, currencyB]) => [currencyA?.wrapped, currencyB?.wrapped]);
   }, [currencies]);
 
-  // const lastBlockNumber = useBlockNumber();
+  const [pairs, setPairs] = useState<[PairState, Pair | null][]>(Array(tokens.length).fill([PairState.LOADING, null]));
+
+  const last = useLastCount(20000);
```
Comment (octocat512) [inline src/hooks/useV2Pairs.ts:30]: any reason to set 20s period?

Comment (lijac) [reply src/hooks/useV2Pairs.ts:30]: The number of API requests was still large when set to 10s. Let's test this more and set a reasonable period.

**Comment by octocat512** [comment]
- [ ] waiting the `getPoolStatsForPair` API upgrade.
- [ ] waiting the multicall score.

**Comment by octocat512** [comment]
Thanks @lijac 
I will merge the PR after review

**Comment by BennyOptions** [comment]
It seems the base/quote is backwards in liquidity details section. It also causes an issue when typing in the text box of IUSDC/bnUSD pair because of decimal differences.

Screen Shot 2022-02-15 at 11 25 24 PM
Screen Shot 2022-02-15 at 11 25 00 PM

**Comment by lijac** [comment]
> It seems the base/quote is backwards in liquidity details section. It also causes an issue when typing in the text box of IUSDC/bnUSD pair because of decimal differences.
> 
> Screen Shot 2022-02-15 at 11 25 24 PM
> 
> Screen Shot 2022-02-15 at 11 25 00 PM

The issue is solved in https://github.com/balancednetwork/balanced-network-interface/pull/880

I've pulled and merged master to this branch. It works now.

**Pull request merged**